### PR TITLE
[TypeChecker] Convert call to subscript into subscript expression

### PR DESCRIPTION
**Pull request by xedin**
When pre-checking expression try to convert call to '.subscript'
into proper subscript expression, such would generate better constraints
and helps to yield better diagnostics in case of failure.

Resolves: <rdar://problem/27017206>.

### CODE REVIEW

**Comment by xedin** [comment]
/cc @rudkx @DougGregor 

This resolves one of the Sema crashers from validation tests, I saw comment in SubscriptExpr::create about not adding new callers, but I think this use-case might be legitimate exception since PreCheckExpression class doesn't have any auxiliary methods to handle argument labels, trailing closures etc.

File: validation-test/Sema/type_checker_crashers_fixed/rdar27017206.swift [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: not --crash %target-swift-frontend %s -parse
+// RUN: %target-swift-frontend %s -parse
 // REQUIRES: asserts
```
Comment (jrose-apple) [inline validation-test/Sema/type_checker_crashers_fixed/rdar27017206.swift]: Drive-by comment: please don't forget to remove this from fixed compiler-crashers test cases.

**Comment by rudkx** [comment]
I don't think we want to do this. We used to reject this, and should probably start rejecting this again unless someone writes a swift-evolution proposal to allow this alternate spelling.

Someone else had a PR to reject this, but I don't know where that ended up.

**Comment by xedin** [comment]
So instead of allowing to call .subscript it should fail with diagnostic saying that '.subscript' can't be called directly please use [] instrad or something like that?

**Comment by rudkx** [comment]
This is the other PR I was thinking of: https://github.com/apple/swift/pull/5537

**Comment by xedin** [comment]
Ah, cool, I will let it be handled there then! Thanks, @rudkx!
